packages/vue-components/src/panels/PanelBase.js

```diff
@@ -95,80 +95,133 @@ export default {
+      this.isRetrieverLoadDone = true;
+
+      if (this.preloadBool && !this.wasRetrieverLoaded) {
+        // Only preload, do not expand the panel.
+        return;
+      }
+      // Don't play the transition for this case as the loading should feel 'instant'.
       if (this.expandedBool) {
         this.$refs.panel.style.maxHeight = 'none';
         return;
       }
 
-      /*
-      Otherwise, since the vue transition is dependent on localExpanded, we have to manually
-      set our own transition end handlers here for the initial loading of the content.
-      */
-      const onExpandDone = () => {
-        this.$refs.panel.style.maxHeight = 'none';
-        this.$refs.panel.removeEventListener('transitionend', onExpandDone);
-      };
-
-      this.$refs.panel.addEventListener('transitionend', onExpandDone);
+      // For expansion transition to 'continue' after src is loaded.
       this.$refs.panel.style.maxHeight = `${this.$refs.panel.scrollHeight}px`;
     },
-    beforeExpand(el) {
-      el.style.maxHeight = '0';
-    },
-    duringExpand(el) {
-      jQuery('html').stop();
-      el.style.maxHeight = `${el.scrollHeight}px`;
-    },
-    afterExpand(el) {
-      el.style.maxHeight = 'none';
-    },
-    beforeCollapse(el) {
-      el.style.maxHeight = `${el.scrollHeight}px`;
+    initCollapsedPanelHeight() {
+      // To set panel's default collapsed panel height.
+      this.collapsedPanelHeight = 0;
     },
-    duringCollapse(el) {
-      if (this.$el.getBoundingClientRect().top < 0) {
-        jQuery('html').animate({
-          scrollTop: window.scrollY + this.$el.getBoundingClientRect().top - 3,
-        }, 500, 'swing');
+    initPanel() {
+      this.$refs.panel.addEventListener('transitionend', () => {
+        /*
+          If the panel's content grows, its height should accomodate it accordingly (e.g. nested panels).
+          We will always set the panel's maxHeight to 'none' after expansion.
+          This is to prevent maxHeight (which is used for transition) from restricting panel's growth.
+
+          At the end of first expansion transition, if src has not finished loading, then we should not
+          set panel's maxHeight to none.
+
+          This is because once src finishes loading, retriever component will emit an event to
+          set the flag isRetrieverLoadDone to true and update the maxHeight again.
+
+          This will trigger a second expansion transition. And when this happens, we need the original
+          maxHeight (from first expansion transition) to ensure the transition can 'continue' smoothly.
+
+          Once that 'second' transition is finished, the flag isRetrieverLoadDone would have already
+          been set to true and the end of the 'second' transition will trigger this event listener
+          again to set maxHeight to none.
+        */
+        if (this.localExpanded && this.isRetrieverLoadDone) {
```

Comment:
Thanks for spotting this! I should have set the maxHeight to none for any panels that were indicated to be expanded. 
However, I was thinking that this event bubbling for transitionend perhaps would be beneficial for our use case? :O 

Since it seems like for any inner nested panel that we expand, the outer panels' maxHeight should always be set to none regardless. It seems like this "feature" would provide sort of a safeguard. Although I agree that this might not be the "right" way to do things, since it may bring in other unexpected behaviours. 

Is this something that we can perhaps keep because it seems to be useful for our use case? Or is it a poor practice that might cause more problems down the road?

Reply:
additional safeguards isn't necessarily bad; It would really depend on the use case / performance considerations / etc. Though over here my preference would be to make it work right from the start, since there's only a finite number of cases to verify

More importantly `transitionend` events are very often emitted. (note not only nested panels - e.g. any element / component that has a fade in transition on load could interfere with the transition)

Reply:
Understood! Thank you for providing your insights for the design decision!